### ISSUE TITLE
performance difference between 5.1 and 6.1 on M1Pro

### ISSUE THREAD

**Issue by Cybergate9** [brycedrennan/imaginAIry#113]
hi again :-)
trying to understand difference between 5.1 and 6.1 performance..
- on M1Pro 8/14, 16GB, OSX 13.0.1, Python 3.10.8, each imaginary installed in its own venv
- generate key parameters same in test - 512x512px seed:xyz prompt-strength:7.5 steps:35 sampler-type:k_dpmpp_2m
- by model hash filename both selecting SD-1.5 as expected.

5.1 consistently generates at 45-50 secs per image
whereas
6.1 consistently generates at 3:45-4:00 minutes per image

is the primary difference here just SD-2.0 code versus SD-1.4 code?

**Comment by brycedrennan** [comment]
This is not expected.

**Comment by Cybergate9** [comment]
a bit more fyi..
So tested 6.1.1, slight apparent improvement.. but gen. times still up to 4ish minutes
On my machine its a memory issue, doing simple monitoring with activity monitor gives memory usage for python while Imaginairy running:
5.1 - using 10-13 GB
6.1.1 - using 12-17GB
So, of course, on a 16GB machine under 6.1.1 it can push swap up to 3-6GB.
So that makes sense that under those conditions performance get murdered.
The proof would be to run on a 32GB M1 machine I suppose.
It's almost as though a model is being loaded (circa 4GB) and not used? 
or a genuine code memory use difference difference between sd1.4 and sd2.0?

**Comment by brycedrennan** [comment]
I had removed some memory optimizations in my rush to get 2.0 out so that seems a likely culprit. I don't experience that same slowdown on my M1 max.  I have a fix in a branch but my test suite isn't passing now so I'm looking into that.

**Comment by Cybergate9** [comment]
ah and it just failed, so maybe something a bit more concrete?
output was:
```
 51%|██████████████████████                     | 18/35 [02:00<01:54,  6.72s/it]
Traceback (most recent call last):
  File "/Users/shaun/dev/py-venv3/bin/imagine", line 8, in <module>
    sys.exit(imagine_cmd())
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/cmds.py", line 258, in imagine_cmd
    imagine_image_files(
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/api.py", line 75, in imagine_image_files
    for result in imagine(
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/api.py", line 309, in imagine
    samples = sampler.sample(
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/samplers/kdiff.py", line 120, in sample
    samples = self.sampler_func(
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/vendored/k_diffusion/sampling.py", line 914, in sample_dpmpp_2m
    denoised = model(x, sigmas[i] * s_in, **extra_args)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/samplers/kdiff.py", line 248, in forward
    noise_pred = get_noise_prediction(
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/samplers/base.py", line 81, in get_noise_prediction
    noise_pred_neutral, noise_pred_positive = denoise_func(
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/samplers/kdiff.py", line 231, in _wrapper
    return self.inner_model(
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/vendored/k_diffusion/external.py", line 130, in forward
    eps = self.get_eps(input * c_in, self.sigma_to_t(sigma), **kwargs)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/vendored/k_diffusion/external.py", line 160, in get_eps
    return self.inner_model.apply_model(*args, **kwargs)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/modules/diffusion/ddpm.py", line 1200, in apply_model
    x_recon = self.model(x_noisy, t, **cond)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/modules/diffusion/ddpm.py", line 1325, in forward
    out = self.diffusion_model(x, t, context=cc)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/modules/diffusion/openaimodel.py", line 836, in forward
    h = module(h, emb, context)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/modules/diffusion/openaimodel.py", line 86, in forward
    x = layer(x, context)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/modules/attention.py", line 441, in forward
    x = block(x, context=context[i])
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/modules/attention.py", line 354, in forward
    return checkpoint_eval(
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/modules/diffusion/util.py", line 149, in checkpoint
    return CheckpointFunction.apply(func, len(inputs), *args)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/modules/diffusion/util.py", line 162, in forward
    output_tensors = ctx.run_function(*ctx.input_tensors)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/modules/attention.py", line 361, in _forward
    self.attn1(
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/modules/attention.py", line 175, in forward
    return self.forward_splitmem(x, context=context, mask=mask)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/modules/attention.py", line 212, in forward_splitmem
    q, k, v = map(
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/imaginairy/modules/attention.py", line 213, in <lambda>
    lambda t: rearrange(t, "b n (h d) -> (b h) n d", h=h), (q_in, k_in, v_in)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/einops/einops.py", line 424, in rearrange
    return reduce(tensor, pattern, reduction='rearrange', **axes_lengths)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/einops/einops.py", line 368, in reduce
    return recipe.apply(tensor)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/einops/einops.py", line 211, in apply
    return backend.reshape(tensor, final_shapes)
  File "/Users/shaun/dev/py-venv3/lib/python3.10/site-packages/einops/_backends.py", line 84, in reshape
    return x.reshape(shape)
RuntimeError: current_allocated_size() >= m_low_watermark_limit INTERNAL ASSERT FAILED at "/Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/mps/MPSAllocator.mm":389, please report a bug to PyTorch.
```

**Comment by brycedrennan** [comment]
yeah that error means there isn't enough memory

**Comment by brycedrennan** [comment]
Unrelated: do you find 35 steps looks better than 15 steps.  I haven't seen a difference in the few tests I've run on that.

**Comment by Cybergate9** [comment]
re memory error - yep, suspected it might tell you no more than that..

re steps - I vary from your default (15) up to 40 steps depending on what I'm doing, sometimes I think there's a difference but its pretty subjective, and I think it depends on sampler too, more subjectivity..

Overall though, re sd2, I'm not particularly bothered to be honest, because the outstanding performance of your 1.5 branch, as well as the output of the 1.5 models seems to be the sweet spot right now.. 

That the thing will run on this chip at that performance is a credit to you!

I will keep looking at sd2 branch but it might turn out its memory requirements are just too high for 8-16gb m1's

**Comment by brycedrennan** [comment]
okay pushed out the fix.

**Comment by brycedrennan** [comment]
Thanks! yeah, sd2 at 768 technically runs on the M1 max 32gb, but it runs super slow in the same way you described here. I assume because of memory issues.

**Comment by idvorkin** [comment]
Hey @brycedrennan - Amazing app 

I'm  on a M1 w/16 GB, when I force the model to 1.4, it still takes ~5 minutes.  What is the command to get the 1 minute  execution  ..

> imagine --model  SD-1.4 "a scenic landscape"

Also how do I see what version of imaginAIry  Im running and  if I have your latest fix?

**Comment by brycedrennan** [comment]
`pip freeze | grep -i imaginairy` should show you the version number

@Cybergate9 can you validate that this was fixed in the latest version or still broken?

**Comment by idvorkin** [comment]
Thanks @brycedrennan  - yup so I'm on 7.0.0, happy to run any diagnostics if that's helpful.

**Comment by brycedrennan** [comment]
@idvorkin you could try running 5.1 with `pip install imaginairy==5.1.0` and see if it runs faster. If so, we'll open this ticket up again.

**Comment by Cybergate9** [comment]
> @Cybergate9 can you validate that this was fixed in the latest version or still broken?

re performance:

I think it's stabilised a lot. It is slower than 1.5 but no longer significantly so, if you compare apples to apples (e.g 512x512)

7.0.0 rough timing tests (fresh reboot, M1Pro 8/14, 16 GB, minimum other apps running, default settings apart from prompt + models, 10 repeats per test, sampling only results in minutes:seconds)

Prompt: "painting of an art deco frog by {_artist_}"

SD-1.5 512x512px seed:random prompt-strength:7.5 steps:15 sampler-type:k_dpmpp_2m
result set 1: median/mean cluster around 1:30 (min:secs), low: 1:00, high 2:00
result set 3: failed with a memory error at image 6/10 (quit terminal to release memory, started again for set 4)
result set 4: median/mean cluster around 1:00 (min:secs), low: 0:40, high 1:49

SD-2.0 512x512px seed:random prompt-strength:7.5 steps:15 sampler-type:k_dpmpp_2m
result set 2: median/mean cluster around 1:20(min:secs), low: 1:00, high 1:31
result set 5: median/mean cluster around 1:25(min:secs), low: 0:40, high 1:43

**Comment by idvorkin** [comment]
OK, trying on my MBP 14" what has the M1 Pro - On 7.0, I'm closer to 50s regardless of SD-1.5 or SD-1.2

**Comment by brycedrennan** [comment]
Fixed in 7.1.1.

Root cause was with the new code the config needed `use_ema: False`

Also I found in testing that generating images runs faster on torch 1.12.1 than on 1.13 for some reason.
